src/system-health/health_checker/manager.py

```diff
@@ -0,0 +1,101 @@
+        """
+        from .service_checker import ServiceChecker
+        from .hardware_checker import HardwareChecker
+        self._checkers.append(ServiceChecker())
+        self._checkers.append(HardwareChecker())
+
+    def check(self, chassis):
+        """
+        Load new configuration if any and perform the system health check for all existing checkers. 
+        :param chassis: A chassis object.
+        :return: A tuple. The first element indicate the status of the checker; the second element is a dictionary that
+        contains the status for all objects that was checked.
+        """
+        from .health_checker import HealthChecker
+        HealthChecker.summary = HealthChecker.STATUS_OK
+        stats = {}
+        self.config.load_config()
+        # check state first to avoid user change boot timeout in configuration file
+        # after finishing system boot
+        if self._state == self.STATE_BOOTING and self._is_system_booting():
+            self._set_system_led(chassis, self.config, 'booting')
+            return self._state, stats
+
+        for checker in self._checkers:
+            self._do_check(checker, stats)
+
+        if self.config.user_defined_checkers:
+            from .user_defined_checker import UserDefinedChecker
+            for udc in self.config.user_defined_checkers:
+                checker = UserDefinedChecker(udc)
+                self._do_check(checker, stats)
+
+        led_status = 'normal' if HealthChecker.summary == HealthChecker.STATUS_OK else 'fault'
+        self._set_system_led(chassis, self.config, led_status)
+
+        return self._state, stats
+
+    def _do_check(self, checker, stats):
+        """
+        Do check for a particular checker and collect the check statistic.
+        :param checker: A checker object.
+        :param stats: Check statistic.
+        :return: 
+        """
+        try:
+            checker.check(self.config)
+            category = checker.get_category()
+            info = checker.get_info()
+            if category not in stats:
+                stats[category] = info
+            else:
+                stats[category].update(info)
+        except Exception as e:
+            from .health_checker import HealthChecker
+            error_msg = 'Failed to perform health check for {} due to exception - {}'.format(checker, repr(e))
+            entry = {str(checker): {
+                    HealthChecker.INFO_FIELD_OBJECT_STATUS: HealthChecker.STATUS_NOT_OK,
+                    HealthChecker.INFO_FIELD_OBJECT_MSG: error_msg
+                }}
+            if 'Internal' not in stats:
+                stats['Internal'] = entry
+            else:
+                stats['Internal'].update(entry)
+
+    def _is_system_booting(self):
+        from .utils import get_uptime
+        uptime = get_uptime()
+        if not self.boot_timeout:
+            self.boot_timeout = self.config.get_bootup_timeout()
+        booting = uptime < self.boot_timeout

```

Comment:
System monitor service is depending on monit service, so it needs to wait monit service to start working. So the boot time out here is actually getting from monit service configuration file located at /etc/monit/monitrc. I also wanted to monitor some expected services and/or containers, but I found that there is no such definition of "boot up done" for all platforms now. So we are using a boot time out to simplify the implementation for now.

Reply:
Even if we eventually have a definition of "boot up done" for all platforms, we will still need some sort of max timeout here, in case the device never reaches "boot up done" for some reason.

To Sujin's point, it would be optimal if we could set this timeout based on the type of boot being performed. One approach I can think of is to have the different timeout values in the config file and choose on based on the type of boot being performed based on the previous reboot cause.

Reply:
As system health service is designed to depend on monit service to check the service/process/file system status, so system health always waits for monit service. I did some tests with monit service and found that monit service always wait 5 minutes after the reboot no matter what the reboot type is, so system health service have to wait 5 minutes here for all reboot type.

Reply:
Understood. For us to use a variable timeout here, we would first need to implement a variable delay for Monit.